nemo/collections/asr/parts/k2/w_transducer.py

```diff
@@ -0,0 +1,337 @@
+        end_eps_id = vocab_size + 1
+        device = units_tensor.device
+        text_len = units_tensor.shape[0]
+
+        # arcs: scr, dest, label, score
+        arcs = torch.zeros(((text_len + 1) * 2 + 2, 4), dtype=torch.int32, device=device)
+        text_indices = torch.arange(0, text_len + 1, dtype=torch.int32, device=device)
+        # eps
+        arcs[0, 2] = start_eps_id
+        # blank labels
+        arcs[1:-1:2, 0] = text_indices  # from state
+        arcs[1:-1:2, 1] = text_indices  # to state
+        arcs[1:-1:2, 2] = blank_id
+
+        # text labels
+        arcs[2:-1:2, 0] = text_indices  # from state
+        arcs[2:-1:2, 1] = text_indices + 1  # to state
+        arcs[2:-2:2, 2] = units_tensor  # labels: text
+
+        arcs[-1] = arcs[-2]
+        arcs[-2, 1] = text_len
+        arcs[-2, 2] = end_eps_id
+        arcs[-1, 2] = -1  # last transition to final state, ilabel=-1 (special for k2)
+        olabels = arcs[:, 2].detach().clone()  # same as ilabels
+
+        fsa_text = k2.Fsa(arcs, olabels)
+        fsa_text.unit_positions = torch.zeros_like(olabels)
+        fsa_text.unit_positions[1:-1] = text_indices.expand(2, -1).transpose(0, 1).flatten()
+        fsa_text.unit_positions[-1] = -1
+        return fsa_text
+
+    def get_temporal_schema(self, num_frames: int, vocab_size: int, device: torch.device) -> "k2.Fsa":
+        """
+        Get temporal schema graph for W-Transducer loss (Compose-Transducer).
+
+        Example graph: blank=0, num_frames=3, vocab_size=3, last_blank_mode="force_last".
+        Labels: <unit>:<frame_index>. <unit> is a unit from vocab + special eps ids `vocab_size`, `vocab_size+1`.
+
+        graph::
+
+                                                         4:0
+                       +--------------------------------------------+
+                       |                               4:1          |
+                       |                     +--------------------+ |
+                1:0    |              1:1    |              1:2   | |
+              +-----+  |            +-----+  |            +-----+ | |
+              v     |  |            v     |  |            v     | v v
+            +--------------+  0:0  +------------+  0:1   +------------+  0:2   +---+  -1:-1   #===#
+            |    0         | ----> |    1       | -----> |    2       | -----> | 3 | -------> H 4 H
+            +--------------+       +------------+        +------------+        +---+          #===#
+              ^ 2:0 | | | |         ^ 2:1 |  ^            ^ 2:2 |  ^             ^
+              +-----+ | | |         +-----+  |            +-----+  |             |
+                      | | |     3:0          |                     |             |
+                      | | +------------------+     3:0             |             |
+                      | +------------------------------------------+    3:0      |
+                      +----------------------------------------------------------+
```

Comment:
The arc from states 0 to 3 is redundant. It will never be used as it will be eaten up during the composition.
It also makes it harder to understand the schema.

Reply:
Removed this arc, also fixed construction code for temporal schema and tests